### ISSUE TITLE
Duplicate and unexpected node chips in k8s Assets view (v1.98)

### ISSUE THREAD

**Issue by dwbrown2** [kubecost/cost-analyzer-helm-chart#1675]
Thanks for your interest in Kubecost!  To ensure you get the help you need, please follow these guidelines:

**Describe the bug**  
Confusing collection on chips show on certain multi-cluster and hosted deployments. For example, the link below shows "No cluster" and then near duplicates of CPU measurements. 

<img width="1013" alt="image" src="https://user-images.githubusercontent.com/298359/187765907-a545193c-ffc2-4546-a8f2-eeb72ae70394.png">

Yesterday, a user showed a hosted deployment when they had many of these repeated and it led to confusion. Looking at backend responses it looks like this is a frontend issue, potentially related to rounding.

**To Reproduce**  
1. Visit https://multi2.test.kubecost.io/assets?agg=type&context=W3sicHJvcGVydHkiOiJTZXJ2aWNlIiwidmFsdWUiOiJLdWJlcm5ldGVzIiwibmFtZSI6Ikt1YmVybmV0ZXMifV0%3D
2. View node chips 

**Expected behavior**  
* Not having the "No cluster" chip given we're looking in k8s
* Not having the duplicated chips that look really similar

**Comment by dwbrown2** [comment]
@Adam-Stack-PM @nealormsbee let me know if I can help share more info for triage

**Comment by Adam-Stack-PM** [comment]
@dwbrown2 Let's get this routed to @mbolt35 as he is on support this week for the triage. 

 Bolt, could you take a deeper dive to verify it is a rounding issue then label with a priority status?

**Comment by mbolt35** [comment]
It also looks like if you drill down into the Node, the problem exists there as well:

![image](https://user-images.githubusercontent.com/334480/187995064-b6920800-ea58-462e-a819-5bc09204d2a4.png)

**Comment by nealormsbee** [comment]
Took a look. The "No Cluster" issue, and the ~similar duplicate chips issue have two different causes.

### Duplicate Chips

Chips for an Asset are determined by looking at the properties of the Asset for each day in the queried range. The, because the table row represents the entire range, a union of the chips from each day is taken. The value of the `GiB RAM` chip is determined using the `ramBytes` property of the Asset, while the `vCPU` chip's value is determined using the `cpuCores` property. If these values are different for the Asset on two different days in the queried range, then we end up with two values for those chips.

I've only looked at the `Node` row from the OP (not the drilldown). In that query response, most of the days have a `ramBytes` value of `166938120192`, however the `2022-09-01 :: 2022-09-02` block has a `ramBytes` value of  `167107344554.66666`. I can't say whether that's a rounding error or some other issue, but the API response is different. You can cURL it with

```
curl 'https://multi.test.kubecost.io/model/assets?window=7d&aggregate=type&accumulate=false' \
  -H 'Accept: */*' \
  -H 'Accept-Language: en-US,en;q=0.9,de;q=0.8' \
  -H 'Cache-Control: no-cache' \
  -H 'Connection: keep-alive' \
  -H 'Origin: http://localhost:3000' \
  -H 'Pragma: no-cache' \
  -H 'Referer: http://localhost:3000/' \
  -H 'Sec-Fetch-Dest: empty' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Site: cross-site' \
  -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/104.0.0.0 Safari/537.36' \
  -H 'sec-ch-ua: "Chromium";v="104", " Not A;Brand";v="99", "Google Chrome";v="104"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "macOS"' \
  --compressed
```

We can fix whatever's causing the data to report differently, but I think we should also come up with a scheme that prevents showing multiple chips for the same property of a row, if that's considered a bug. I don't know if it makes sense to show max ramBytes consumed over the interval, or an average, or something else. It depends on what the chips are meant to indicate.


### No Cluster

We determine this chip by checking the Asset's property `Asset.properties.cluster`. That field seems to be absent in the response, and my guess is that this may be the norm in multi-cluster setups. Guidance on how to handle this appreciated. Should we just not show the chip if there's no `cluster` property? Or should we show `No Cluster` when the property is absent, *unless* we detect that we're in a multicluster environment? Or something else?

**Comment by mbolt35** [comment]
@nikovacevic Would love your opinion here, I feel like you always have good ideas surrounding user facing UI anomalies stemming from ETL data.

**Comment by dwbrown2** [comment]
re: "No Cluster" -- would be interesting to learn why that is missing. @mbolt35 have a sense here? seems like a potential fix for backend. If it's a k8s asset but has "No Cluster" I would consider hiding or showing "Unknown Cluster."  I don't think the current copy is helpful to users. 

re: duplicate changes -- seems like it would also be interesting to understand why backend returns different values as these _should_ be immutable in most cases. I would recommend using the latest known value and only showing one per resource.

@nealormsbee let me know if you disagree with this approach on the FE!

**Comment by mbolt35** [comment]
@dwbrown2 I agree, the "No Cluster" seems like something backend related, but I know that we have some limitations to what we  merge between Assets and Allocations. I'll look into it, although I feel like @nikovacevic may just know the answer off-hand

**Comment by Adam-Stack-PM** [comment]
@dwbrown2, and/other, Love the discussion here.  Would also be helpful to get clear on if this needs to be rushed into v1.97? 

We have lots of items in motion for v1.97 and would love to make sure we don't loose ground triaging items that can be pushed out.

**Comment by dwbrown2** [comment]
I'd call between P1/P2 but caused this user to question data accuracy. Seems like it would be a quick fix on FE. Can wait until 1.98 if needed.

**Comment by nealormsbee** [comment]
Looks like we'll defer this until v1.98, so taking it out of my attention queue for now.

Suggested approach sounds reasonable to me, Webb. If we do happen to get a bad 1d block for an Asset, then the chip will only be wrong for a day if we're using most recent value :+1:

**Comment by dwbrown2** [comment]
@burkeshartsis looks like this is fixed in new UI! That sound right?

**Comment by nealormsbee** [comment]
Yes! This is fixed for the old and new Assets pages in v1.98, using Sets to ensure we don't get multiple disagreeing values for hardware specs. My mistake not tying the fix commit to the issue! Closing.
